src/Featurizers/LagLeadOperatorFeaturizer.h

```diff
@@ -0,0 +1,224 @@
+        static_cast<size_t>(_max_future_offset-_max_prior_offset))
+    ) {
+        // prepopulate circular buffer with null values to imitate non-existing prior rows
+        for (std::int64_t i = 0; i < (static_cast<std::int64_t>(_horizon) - _max_prior_offset - 1); ++i) {
+            _buffer.push(Traits<nonstd::optional<T>>::CreateNullValue());
+        }
+}
+
+template <typename T>
+LagLeadOperatorTransformer<T>::LagLeadOperatorTransformer(Archive &ar) :
+    LagLeadOperatorTransformer(
+        [&ar](void) {
+            // Version
+            std::uint16_t                   majorVersion(Traits<std::uint16_t>::deserialize(ar));
+            std::uint16_t                   minorVersion(Traits<std::uint16_t>::deserialize(ar));
+
+            if(majorVersion != 1 || minorVersion != 0)
+                throw std::runtime_error("Unsupported archive version");
+
+            // Data
+            std::uint32_t horizon = Traits<std::uint32_t>::deserialize(ar);
+            std::vector<std::int64_t> offsets = Traits<std::vector<std::int64_t>>::deserialize(ar);
+            
+            return LagLeadOperatorTransformer(std::move(horizon), std::move(offsets));
+        }()
+    ) {
+}
+
+template <typename T>
+bool LagLeadOperatorTransformer<T>::operator==(LagLeadOperatorTransformer const &other) const {
+    return this->_horizon == other._horizon
+            && this->_offsets == other._offsets;
+}
+
+template <typename T>
+bool LagLeadOperatorTransformer<T>::operator!=(LagLeadOperatorTransformer const &other) const {
+    return (*this == other) == false;
+}
+
+template <typename T>
+void LagLeadOperatorTransformer<T>::save(Archive &ar) const /*override*/ {
+    // Version
+    Traits<std::uint16_t>::serialize(ar, 1); // Major
+    Traits<std::uint16_t>::serialize(ar, 0); // Minor
+
+    // Data
+    Traits<std::uint32_t>::serialize(ar, _horizon);
+    Traits<std::vector<std::int64_t>>::serialize(ar, _offsets);
+
+    // Note that we aren't serializing working state
+}
+
+template <typename T>
+void LagLeadOperatorTransformer<T>::execute_helper(nonstd::optional<typename BaseType::InputType> const &input, typename BaseType::CallbackFunction const &callback) {
+    _buffer.push(input);
+    // until the circular buffer is full, we don't have enough data to generate output matrix
+    if (_buffer.is_full()) {
+        Microsoft::Featurizer::RowMajMatrix<nonstd::optional<T>> ret(_offsets.size(), _horizon);
+        for (std::uint32_t row = 0; row < _offsets.size(); ++row) {
+            // the circular buffer starts with the needed data for the _max_prior_offset
+            // so the difference between offsets[row] and _max_prior_offset is the position to find the data to copy
+            std::tuple<typename Components::CircularBuffer<TargetType>::iterator, typename Components::CircularBuffer<TargetType>::iterator> range = _buffer.range(_horizon, static_cast<size_t>(_offsets[row] - _max_prior_offset));
```

Comment:
>_offsets[row] - _max_prior_offset [](start = 193, length = 33)

Rather than performing this calculation on every call, would it be possible to do it when assigning offsets in the constructor? That may eliminate the need to cache `_max_prior_offset` as well.

Reply:
I think moving the calculation to the constructor is a good idea so instead of keeping the real lag lead offsets, I can just cache relative difference according to `_max_prior_offset`. But I still need to cache `_max_prior_offset` in this approach, right?